libs/ml/include/ml/optimisation/adagrad_optimiser.hpp

```diff
@@ -40,10 +40,11 @@ class AdaGradOptimiser : public Optimiser<T, C>
   using DataType  = typename ArrayType::Type;
   using SizeType  = typename ArrayType::SizeType;
 
-  AdaGradOptimiser(std::shared_ptr<Graph<T>> graph, std::string const &input_node_name,
-                   std::string const &output_node_name,
-                   DataType const &   learning_rate = DataType{0.001f},
-                   DataType const &   epsilon       = DataType{1e-8f});
+  AdaGradOptimiser(std::shared_ptr<Graph<T>>       graph,
+                   std::vector<std::string> const &input_node_names,
+                   std::string const &             output_node_name,
+                   DataType const &                learning_rate = DataType{0.001f},
+                   DataType const &                epsilon       = DataType{1e-8f});
```

Comment:
is there a need to pass learning rate and epsilon by reference? For primitives we should prefer pass by value.
see (https://stackoverflow.com/questions/14013139/is-it-counter-productive-to-pass-primitive-types-by-reference)

Reply:
It makes sense in the case of Tensor<FixedPoint>.